### DISCUSSION TITLE
Berry: Getting individual sensor values to variables

### DISCUSSION THREAD

**Discussion by boozeman** [arendst/Tasmota#13975] [Q&A] [answered]
With 10.1.0.1 Berry Console I can command

`somevar = tasmota.read_sensors()` To get various sensor values to variable.
```
somevar
{"Time":"2021-12-09T17:49:37","ANALOG":{"A1":232},"BME280": {"Temperature":23.1,"Humidity":21.2,"DewPoint":-0.3,"Pressure":1013.6},"SHT3X":{"Temperature":20.0,"Humidity":25.0,"DewPoint":-0.7},"ESP32":{"Temperature":53.3},"PressureUnit":"hPa","TempUnit":"C"}
```
I want to get BME280 Pressure, SHT3X humidity and temperature to variables (for calculation of Equivalent sea level pressure, Absolute Humidity, Vapour Pressure Deficit, etc)

Do I have to mangle that somevar with example `string.split(somevar,"whatever you want to use split")` 
Or is there any easier way to get those single sensor readings out? 

Next step is to add extra values on the SENSOR or equivalent MQTT payload, but that's different story ;-)

For now those extras are generated with Home Assistant templating.

**Comment by sfromis** [comment, marked as answer]
Certainly no need for splitting....
```
import json
sensors=json.load(tasmota.read_sensors())
print(sensors['BME280']['Temperature'])
```

To add your own sensor data to the usual places for sensors (including MQTT), you can create a driver in berry, where you can have methods to create JSON data in exactly the same way as built-in sensors.

Note: If you set [Altitude](https://tasmota.github.io/docs/Commands/#altitude), Tasmota will automatically include SeaPressure when the sensor delivers Pressure.

**Reply by boozeman** [reply]
Thanks for that!

For the note if anybody interested those weather related formulas. They're not perfect but somewhat accurate now. I don't even know that automated pressure set based by elevation.  My goal is to make automated humidity/temperature control for greenhouse with Vapour Pressure Deficit optimal for chilis

```
import json
import math


sensors=json.load(tasmota.read_sensors())
p = sensors['BME280']['Pressure']
rh = sensors['SHT3X']['Humidity']
t = sensors['SHT3X']['Temperature']

# Equivalent sea level pressure - variable ap
elevation = 92
ap = (p / math.pow(1 - ((0.0065 * elevation) / (t + (0.0065 * elevation) + 273.15)), 5.257) / 100); # in hPa

# Absolute Humidity in g/m3
mw = 18.016 # molar mass of water g/mol  
r = 8314.3 # Universal gas constant J/mol/K
ah = (6.112 * math.pow(2.718281828, (17.67 * t) / (t + 243.5)) * rh * mw) / ((273.15 + t) * r) 

# Saturated Vapour Pressure (Pa)
SVP = 610.7 * (math.pow(10, (7.5*t/(237.3+t)))) 

# Actual Vapour Pressure in kPa
VPactual = (rh * SVP/1000)/100.0 

# Vapour Pressure Deficit in kPa
VPD = ((1 - (rh/100))* SVP)/1000 

# Humidity Deficit of the air g/m3
PRT = SVP / (8.31 * (273.15 + t)) 

# Humidity Defict of the air at 0% RH, Water mol mass mw is 18.016
HD = PRT * mw 
HD = HD - (HD * rh/100)
```

**Reply by sfromis** [reply]
Since altitude is among the Tasmota configuration options, you could also do something like:
`elevation = tasmota.cmd("Altitude")['Altitude']`
Of course, doing this at the same place as the calculations would issue the command every time, with associated log/MQTT activity.

**Reply by boozeman** [reply]
Hi again,

I dare to ask one more thing about the drivers. I managed to spill out somewhat working driver for my needs With the help of Tasmota Berry Cookbook's example.
 
```
def greenhouse()
  ..
  ..
  self.greenhouse_data = [ap,ah*1000,VPD,HD]
  return self.greenhouse_data
end

```
function returns greenhouse_data as array like [1016.78, 5.31033,2486.47, 13.0116] if it manages to get temp, pressure and humidity readings from the sensors I use.

With web_sensor() function it works like charm

```
  def every_second()
	if !self.greenhouse return nil end
	self.greenhouse()
  end

  def web_sensor()
    import string
    if !self.greenhouse return nil end	
    var msg = string.format(
                     "{s}SeaPressure2{m}%.3f kPa{e}"..
                     "{s}AbsoluteHumidity{m}%.1f g/m3{e}"..
                     "{s}VPD{m}%.3f kPa{e}"..
                     "{s}Humidity Deficit{m}%.3f g/m3{e}",
                     self.greenhouse_data[0],self.greenhouse_data[1],self.greenhouse_data[2],self.greenhouse_data[3])
    tasmota.web_send_decimal(msg)
  end

```
![image](https://user-images.githubusercontent.com/40159019/145667574-e6a035dc-7bce-4eb3-9591-96fa1cd4db0f.png)

but when I add json_append() to driver

```
  def json_append()
     if !self.greenhouse return nil end
	import string
	var msg = string.format(",\"GreenhouseData\":{\"SeaPressure2\":%i,\"AbsoluteHumidity\":%i,\"VPD\":%i,\"Humidity Deficit\":%i}",
                         self.greenhouse_data[0],self.greenhouse_data[1],self.greenhouse_data[2],self.greenhouse_data[3])
    tasmota.response_append(msg)
  end
```
I got this error:
![image](https://user-images.githubusercontent.com/40159019/145667137-d4001e3d-e0be-4907-93dd-f98c90a320f5.png)

I am sure that You can easily point out why I am so blind with this.

**Reply by barbudor** [reply]
Wrong test in json_append 1st line
self.greenhouse instead of self.greenhouse_data

**Reply by sfromis** [reply]
If you really mean self.greenhouse (as in the other functions) instead of self.greenhouse_data, you do not show anything about how variables are defined and controlled. And you do not show the code where it happens either. It is not really feasible to "remote debug" with most of the relevant info not present.

**Reply by boozeman** [reply]
And also wrong datatype for msg I assume %i is for Integer and %s for string. With these changes it works. 

For next read The [Berry Script Language Reference Manual](https://github.com/berry-lang/berry/wiki/Reference) and try to learn something :-)

**Reply by barbudor** [reply]
Obviously `self.greenhouse` is the function that we see on 1st code snippet. So it always exist and the test doesn't returns `nil`.
Of course you can't have both a method and a variable with the same name.

And I susbscribe to @sfromis comment: small code snippets like this does not help much to understand the whole thing. Better to post more relevant code and the surroundings.
To avoid posting lengthy code into the discussion if you haven't commited it to github yet, you can use gist.github.com or pastebin.com

**Reply by boozeman** [reply]
You are both right for this. 

I managed to get it work. there's the link to gist [GhData.be](https://gist.github.com/boozeman/e191e01ef914e2651544a03a7acf32eb)

Thanks for your help.

**Reply by s-hadinger** [reply]
Thanks for your contribution. Actually your class does not need to derive from `Driver` anymore. You can create a plain class instead with no superclass.

**Reply by barbudor** [reply]
I wonder if by doing this:
```
GhData = GhData()
```
you are not erasing the class itself.

@s-hadinger what's your opinion ?

**Reply by sfromis** [reply]
As I see it, that assignment does not "erase the class", but replacing the variable content it with an instance reference means that you can no longer directly refer to the class as such. With the class still in scope, no real loss. 

You could also assign GhData to another variable and still have the class reference there, as name of the variable holding a class reference does not matter, outside of the circumstance that class creation by default also creates a variable with the same name as the class.

**Reply by s-hadinger** [reply]
If it's meant to create a singleton then it is a good way to do it. Although generally speaking I wouldn't recommended it, this is ok for a singleton. 

The class still exists but is no more accessible through a global name. You can still get it with `classof(GhData)`
